Bluetooth: Host: settings without snprintk use

cvinayak opened the pull request:
Update settings implementation for not reusing the OS snprintk implementation with CONFIG_PRINTK=n.

With PR ​

```
west build -p -d build/peripheral_hids_nrf54l15dk -b nrf54l15dk/nrf54l15/cpuapp samples/bluetooth/peripheral_hids -DCONFIG_BOOT_BANNER=n -DCONFIG_PRINTK=n -DCONFIG_BT_SETTINGS_USE_PRINTK=n -DCONFIG_LOG=n -DCONFIG_OUTPUT_DISASSEMBLY=y
```


```
Memory region         Used Size  Region Size  %age Used
           FLASH:      155012 B      1428 KB     10.60%
             RAM:       27420 B       188 KB     14.24%
        IDT_LIST:          0 GB        32 KB      0.00%
```


Without PR, base branch https://github.com/zephyrproject-rtos/zephyr/tree/0b52928625b06b6b8de9b56524fbc1ae5f8cbd6a

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      157012 B      1428 KB     10.74%
             RAM:       27420 B       188 KB     14.24%
        IDT_LIST:          0 GB        32 KB      0.00%
```
​

[0001-samples-Bluetooth-peripheral_hids-Reduce-code-size-w.patch](https://github.com/user-attachments/files/24770057/0001-samples-Bluetooth-peripheral_hids-Reduce-code-size-w.patch)

20260121:
```
-- Zephyr version: 4.3.99 (/home/vich/workspace/zephyrproject/zephyr), build: v4.3.0-2228-g3aa46ff1adea
[387/387] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      154912 B      1428 KB     10.59%
             RAM:       27420 B       188 KB     14.24%
        IDT_LIST:          0 GB        32 KB      0.00%
```

cvinayak committed:
Bluetooth: Host: settings without snprintk use

Update settings implementation for not reusing the OS
snprintk implementation with CONFIG_PRINTK=n.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Host: Rework settings without snprintk use

Rework settings implementation for not reusing the OS
snprintk implementation with CONFIG_PRINTK=n.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Host: Rework settings without snprintk to return error

Rework settings implementation for not reusing the OS
snprintk implementation with CONFIG_PRINTK=n and to return
error if path buffer is insufficient.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Host: ATT: Do not use bt_addr_le_to_str when CONFIG_LOG=n

Do not use bt_addr_le_to_str() when CONFIG_LOG=n.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -324,21 +374,23 @@ __weak void bt_testing_settings_delete_hook(const char *key)
-		if (err < 0) {
-			return -EINVAL;
-		}
+		err = bt_settings_encode_key_no_addr(key_str, sizeof(key_str), key);
```
I'm OK with 2 functions, but an alternative could also be to just have a single function that takes an `addr` which could be NULL and handle it there

cvinayak replied:
I did push a draft in this PR with both implementation, and the code size with ~ 20 bytes more due to added branching instructions to check for NULL, and to keep old behavior of optional parameters and the string output on truncation etc.

Thalley replied:
Interesting. You'd have the same amount of branching, so not sure why it would be 20 bytes more due to that

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -100,8 +117,41 @@ void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
+		/* Key format:
+		 *  "bt/<key>", "/<key>" is optional
+		 */
+		strcpy(path, "bt/");
```
We should consider putting `"bt/"` into a macro or something; it's hardcoded so many places :(

cvinayak replied:
Was my first thought too... but did not want to deviate too much from the scope and purpose of the PR.

Thalley replied:
Yeah, that change is better for a different PR IMO

cvinayak commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,28 +33,45 @@
+	err = snprintk(path, path_size, "bt/%s", key);
+	if (err < 0) {
+		return -EINVAL;
+	}
```
This the old behavior.
https://github.com/zephyrproject-rtos/zephyr/pull/100887/changes#diff-c0d733611772ea4600bb50136c8a70402314c69f6e20ef8f245d482382a2917fL339-L341

cvinayak commented on subsys/bluetooth/host/settings.c:
```diff
@@ -100,8 +117,41 @@ void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
+	size_t len = 3;
+
+	/* Skip if path_size is less than 3; strlen("bt/") */
+	if (len < path_size) {
```
Kept the old behavior. Do not mind fixing if a suggestion is provided, and the purpose of the PR is maintained.

Thalley replied:
Can you elaborate? The old behavior was `err = snprintk(key_str, sizeof(key_str), "bt/%s", key);`, so not sure how that relates to this.

cvinayak replied:
Added BT_ASSERT

cvinayak commented on subsys/bluetooth/host/settings.c:
```diff
@@ -100,8 +117,41 @@ void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
+		if (len >= path_size) {
+			/* Truncate string */
+			path[path_size - 1] = '\0';
+		}
```
Agree... will need to take this up as separate commit or PR tho.

Thalley replied:
Why should that be in a separate PR, when the offending code is being added by this PR? 

I know the code is heavily inspired by `bt_settings_encode_key`, but I don't see any reasons to copy and keep the bugs from that (the bugs in `bt_settings_encode_key` could even be fixed as part of this PR instead)

cvinayak replied:
The implementation here is trying to meet: `The functions snprintf() and vsnprintf() write at most size bytes (including the terminating null byte ('\0')) to str.`  definition of these function.

cvinayak replied:
The code only sets `path[path_size - 1] = '\0'`  if `strlen(path)`   is equal to `path_size`... the length is not going to exceed `path_size`  as `strncpy` is used.

cvinayak reviewed (commented):
The scope in this PR being:
1. settings to truely respect CONFIG_BT_SETTINGS_USE_PRINTK and not use `snprintk`
  - Original PR: https://github.com/zephyrproject-rtos/zephyr/pull/19208
2. why is `snprintk` and its caller code (bt_addr_le_to_str) not eliminated when `CONFIG_LOG=n` ?

cvinayak commented on subsys/bluetooth/host/att.c:
```diff
@@ -385,8 +384,13 @@ static void att_disconnect(struct bt_att_chan *chan)
 
-	bt_addr_le_to_str(bt_conn_get_dst(chan->att->conn), addr, sizeof(addr));
-	LOG_DBG("ATT disconnecting device %s", addr);
+	if (IS_ENABLED(CONFIG_LOG)) {
```
Use of `va_list`  in `vsnprintk` ?

<img width="867" height="401" alt="image" src="https://github.com/user-attachments/assets/69a2dbac-7a3e-473b-b7ef-c827ea2679b2" />

cvinayak commented:
@wopu-ot Original contribution from Oticon, will be good to get some of your reviewers :-)

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -100,8 +117,41 @@ void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
+	} else if (path_size > 0) {
+		/* path_size not sufficient for "bt/" */
+		*path = '\0';
+	}
```
Should this be an assert? Seems like a programming issue, rather than a settings-content issue

cvinayak commented on samples/bluetooth/peripheral_hids/src/main.c:
Not directly, but the architecture port stating "almost a requirement" to send the console (`printk`, `printf`) output over serial line or RAM console.
https://docs.zephyrproject.org/latest/hardware/porting/arch.html#console-over-serial-line

Thalley replied:
Isn't that just a matter of configuration of the logging system? I don't read the above as a requirement that `printk` shall be supported, and `LOG` shall not.

cvinayak commented on subsys/bluetooth/host/settings.c:
```diff
@@ -100,8 +117,41 @@ void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
+	if (len < path_size) {
+		/* Key format:
+		 *  "bt/<key>", "/<key>" is optional
+		 */
```
i mean, replacement of `snprintk`  use for `bt_settings_encode_key()` has check for `key`  being `NULL`  and `bt_settings_encode_key()` stated `key` can be optional.

cvinayak commented on subsys/bluetooth/host/settings.c:
```diff
@@ -100,8 +117,41 @@ void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
+
+		/* If path string is full, always null terminate at path_size */
+		if (len >= path_size) {
+			/* Truncate string */
```
`strncpy` will not copy memory (and already has check to ensure so, that `path_size - len > 0`).

> Additionally, keys cannot be any size, so there should also be a separate key_len check, right, to ensure it's between 7 and 16 octets when decoded (or is that done somewhere else?)

That should be in SMP?

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +33,127 @@
+	}
+
+	/* Concatenate addr as much the free space permits */
+	for (int8_t i = 5; i >= 0 && len < path_size; i--) {
```
Why `int8_t`?

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +33,127 @@
-				       path_size - len);
-		}
+	/* Postfix '/' and concatenate key as much the free space permits */
+	if (key && len < path_size) {
```
```suggestion
	if (key != NULL && len < path_size) {
```

cvinayak replied:
done

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +33,127 @@
+	if (len >= path_size) {
+		/* Truncate string */
+		path[path_size - 1] = '\0';
+	}
```
There is where I think we need to do something different. If the buffer isn't large enough to hold the content, we shouldn't just silently truncate it

cvinayak replied:
Should I replace with `BT_ASSERT(len < path_size)` ?

Thalley replied:
Something like that. I really think that we shouldn't silently truncate data like this. Not sure if a runtime error (returning an error to the settings subsystem) or an assert is best

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -324,21 +376,23 @@ __weak void bt_testing_settings_delete_hook(const char *key)
-	int err;
-	char id_str[4];
 	char key_str[BT_SETTINGS_KEY_MAX];
+	char id_str[4];
```
You can move the `id_str` to the `if (addr)` clause to reduce scope

cvinayak commented on samples/bluetooth/peripheral_hids/src/main.c:
```diff
@@ -37,45 +37,54 @@ static const struct bt_data sd[] = {
-	printk("Connected %s\n", addr);
-
 	if (bt_conn_set_security(conn, BT_SECURITY_L2)) {
 		printk("Failed to set security\n");
```
Only implementation with code reduction opportunity is conditional compiled.

cvinayak commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +33,127 @@
-			path[len] = '/';
-			len++;
-		}
+		path[len] = '0' + addr->type;
```
`0` (immediate value upto 5 bits) are part of ARM thumb instruction and do not need extra code space. With  and without `hex2char` use the code space is same, but increased CPU cycle due to branching (instruction pipeline being flushed, may be branch prediction in CPUs helps?).

Assignment statement is:
```
 8355                 path[len] = '0' + addr->type;                                                       
 8356     330e:       782b            ldrb    r3, [r5, #0]                                                
 8357     3310:       3330            adds    r3, #48 ; 0x30                                              
 8358     3312:       553b            strb    r3, [r7, r4]                                                
```

With `hex2char` reused (as used in `bin2hex`):
```
 8355                 hex2char(addr->type, &path[len]);                                                   
 8356     330c:       f898 0000       ldrb.w  r0, [r8]                                                    
 8357     3310:       1931            adds    r1, r6, r4                                                  
 8358     3312:       f018 fcc4       bl      1bc9e <hex2char>                                            
```

Thalley replied:
> Original contribution was a non-ARM architecture, hence, atleast two architectures are efficient without branching in this scenario.

Assuming that the original author actually tested and/or considered this :) 

> As seen from builds, it is significant code space increase (format strings for snprintk take code space in addition to snprintk 's own code use). Simple specific concatenation save on the heavy caller and callee code use by snprintk. It may be balanced if there are more module specific concatenations. For a resource constraints device, users be given the option to optimize application with less forking of the subsystems.

That sort of brings us back to what I think Johan or Aleksander said the other day: Why do we even have both versions then if the non-snprintk version is always smaller? 

I guess if it's smaller here, then `bt_addr_to_str` would probably also benefit from non using `snprintk` (along with a variety of other functions used by BT), but that is of course outside the scope of this PR

cvinayak replied:
As pointed out early `bt_add_to_str` is not simple concatenation of string buffer, but a mix of constants and variable meeting a format; best handled by `snprintk`. Additionally, use of `bt_addr_to_str` is not a functional requirement for most constraint devices with no user interface.

cvinayak commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +33,127 @@
+	for (int8_t i = 5; i >= 0 && len < path_size; i--) {
+		len += bin2hex(&addr->a.val[i], 1, &path[len],
+			       path_size - len);
+	}
```
> Question: if the application uses `snprintk` already for something, does disabling `CONFIG_BT_SETTINGS_USE_PRINTK` keep the same code size, or does it increase?

Code size will increase marginally, 124 bytes.

`DCONFIG_PRINTK=y -DCONFIG_BT_SETTINGS_USE_PRINTK=n`  ~5KB increase.

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      160076 B      1428 KB     10.95%
             RAM:       27428 B       188 KB     14.25%
        IDT_LIST:          0 GB        32 KB      0.00%
```

`-DCONFIG_PRINTK=y -DCONFIG_BT_SETTINGS_USE_PRINTK=y` ~5KB increase. Reusing `snprintk`  saves 124  bytes. 

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      159952 B      1428 KB     10.94%
             RAM:       27428 B       188 KB     14.25%
        IDT_LIST:          0 GB        32 KB      0.00%
```

`-DCONFIG_PRINTK=n -DCONFIG_BT_SETTINGS_USE_PRINTK=n` ~ 2KB less compare to main.

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      155012 B      1428 KB     10.60%
             RAM:       27420 B       188 KB     14.24%
        IDT_LIST:          0 GB        32 KB      0.00%
```

`-DCONFIG_PRINTK=n -DCONFIG_BT_SETTINGS_USE_PRINTK=n` on base branch:

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      157012 B      1428 KB     10.74%
             RAM:       27420 B       188 KB     14.24%
        IDT_LIST:          0 GB        32 KB      0.00%
```

Thalley replied:
> Code size will increase marginally, 124 bytes.

So the optimization done here may increase flash if the application uses printk for anything else, including something as trivial as logging addresses with `bt_addr_to_str` which is quite commonly used. So this optimization (if enabled) will by default likely increase memory usage, and only in the case of `-DCONFIG_PRINTK=n -DCONFIG_BT_SETTINGS_USE_PRINTK=n` will provide an actual optimization, but by disabling features that used by subsystems, and may even prevent certain features to be enabled? 

For example, HAS uses `bt_addr_le_str` in `LOG_ERR`, and it would not be possible to have logging and HAS enabled with `CONFIG_PRINTK=n`, in which case the memory footprint from this PR will increase when `CONFIG_BT_SETTINGS_USE_PRINTK=n`. 

It seems to me that as long as Bluetooth uses `snprintk` (or any printk for that matter) as much as it does, then the optimization from this PR is the opposite of an optimization, and is rather increased code to maintain with a higher memory footprint. 

It looks to me that BT either needs to lean into `printk` or away from it. Optimizing a single (or 2) function like this won't do us any good IMO and I think it's actively harmful in the majority of cases.

Thalley replied:
Fair points :) Ideally it'd be an optimization in any case, which would make it possible to have it as the only function, but it is what is it :)

Mesh and Classic depend on printk today regardless, and the API functions `bt_uuid_to_str` and `bt_addr_to_str` does too, so possible `CONFIG_BT_SETTINGS_USE_PRINTK` should depend on neither Classic nor Mesh being enabled (as that would just be a bad choice (or maybe just be a warning?)), and then we may want to consider if `bt_addr_to_str` and `bt_uuid_to_str` should either be changed to not depend on printk, or have 2 versions like these functions here

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +33,129 @@
+
+	/* Concatenate key as much the free space permits */
+	strncpy(&path[len], key, path_size - len);
+	len = strlen(path);
```
I guess it's technically possible to use the return value of `strncpy` to determine if `len >= path_size` without calling `strlen` for a minor optimization (making it an O(1) check rather than a O(n)). Something like 
```c
end = strncpy(...)
if (end - path >= path_size) { ... }
```

It's OK as is (although I'd still prefer the length checks at the beginning of the function before attempting to copy anything)

cvinayak commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +33,129 @@
+		path[len] = '/';
+		len++;
+		strncpy(&path[len], key, path_size - len);
+		len += strlen(&path[len]);
```
Good catch.

From the `strncpy` man pages:

`It's impossible to distinguish truncation by the result of the call, from a character sequence that just fits the destination buffer; truncation should be detected by comparing the length of the input string with the size of the destination buffer.` 

Hence updated to use `min(strlen(key), (path_size - len))`

cvinayak commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +33,129 @@
+		path[len] = '/';
+		len++;
+		strncpy(&path[len], key, path_size - len);
+		len += strlen(&path[len]);
```
To save on CPU cycles counting entire string, accumulating only the concatenated new length.

cvinayak committed:
Bluetooth: Host: settings: Fix truncation detection after strncpy

Fix truncation detection after strncpy, as strlen cannot be
used if the src was truncated due to insufficient size of
dst.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +33,129 @@
+	/* path_size is less than or equal 3; strlen("bt/") */
+	if (path_size <= len) {
+		return -EINVAL;
+	}
```
Personally I would have written the function like

```c
int bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
			   const bt_addr_le_t *addr, const char *key)
{
	/* Key format:
	 *  "bt/<subsys>/<addr><type>/<key>", "/<key>" is optional
	 */
	const size_t prefix_len = sizeof("bt") - 1;
	const size_t subsys_len = sizeof('/') + strlen(subsys);
	const size_t addr_str_len = sizeof('/') + sizeof("554433221100tt") - 1;
	const size_t key_len = key == NULL ? 0 : sizeof('/') + strlen(key);
	const size_t total_len = prefix_len + subsys_len + addr_str_len + key_len + sizeof('\0');
	size_t offset = 0U;

	if (path_size < total_len) {
		return -EINVAL;
	}

	memcpy(path, "bt", prefix_len);
	offset += prefix_len;

	path[offset+] = '/';
	memcpy(path + offset, subsys, subsys_len);
	offset += subsys_len;

	path[offset+] = '/';
	for (int8_t i = 5; i >= 0 && len < path_size; i--) {
		hex2char(&addr->a.val[i], &path[offset++]);
	}
	hex2char(&addr->type, &path[offset++]);

	if (key != NULL) {
		path[offset+] = '/';
		memcpy(path + offset, key, key_len);
		offset += key_len;
	}
	path[offset] = '\0';

	LOG_DBG("Encoded path %s", path);

	return 0;
}
```

I don't have any specific numbers on this, but I would expect this to not only be smaller in code size, but also more performant as `strcpy` and `strncpy` not only contains additional checks for each character copies, but each call also stores the NULL terminator. Getting the lengths and doing the length check at the beginning should not affect performance in such a high degree that your argument that it's unlikely. 

However what is the most performant solution also likely depends on the compiler, the platform/architecture and the compiler options. 

It would also be possible to use something like a net_buf_simple to increment the offset automatically instead, but not sure what the complications of performance of that would be. 

In any case, I think the performance differences we discuss here are so minimal that it doesn't really matter for something that's only done once per boot (probably). What is more important (to me) is correctness (which the current implementation (and even the version of this PR right now) struggles with (we should not truncate data...), and readability. It's not particularly clear why we have things like `len += MIN(strlen(subsys), (path_size - len));` after `strncpy`s IMO. 

I won't be blocking the PR once the truncations statements and code have been removed, as the performance metrics are quite hard to draw specific conclusions from IMO.

cvinayak replied:
> Personally I would have written the function like

That is 20 extra bytes on the program stack which cannot be detected at build time, needing runtime thread analyze to update stack size requirements of the applications.

Also, memcpy with length supplied is inefficient with more code executing to check overlapping memory, word alignments and length counting compared to strcpy which terminates on null.

example:
```
00012c3c <strcpy>:
   12c3c:  4603        mov  r3, r0
   12c3e:  f811 2b01   ldrb.w  r2, [r1], #1
   12c42:  f803 2b01   strb.w  r2, [r3], #1
   12c46:  2a00        cmp  r2, #0
   12c48:  d1f9        bne.n  12c3e <strcpy+0x2>
   12c4a:  4770        bx  lr

```

> It's not particularly clear why we have things like len += MIN(strlen(subsys), (path_size - len)); after strncpys IMO.

This is the suggestion in the `strncpy` man pages to handle truncation scenarios. `truncation should be detected by comparing the  length  of  the  input string with the size of the destination buffer.` I am sure the `strncpy` documentation has been thoroughly reviewed for this being mentioned there.

Thalley reviewed (dismissed):
I don't really want to discuss the optimizations anymore, as they are so minimal that it hardly matters IMO :) The only thing I want to have fixed now is proper checks for the length so that we don't silently truncate data

cvinayak commented:
> I don't really want to discuss the optimizations anymore, as they are so minimal that it hardly matters IMO :) The only thing I want to have fixed now is proper checks for the length so that we don't silently truncate data

There is the late check and hence we don't silently truncate data. Your suggestion for readability is not justified with significant increase to runtime RAM and CPU usage, potentially increasing risk of failures.

cvinayak committed:
Bluetooth: Host: settings: Add early len check without strncpy/cmp use

Add early len check, remove use of strcpy, strncpy and
strncmp.

Co-authored-by: Emil Gydesen <emil.gydesen@nordicsemi.no>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +32,122 @@
+	path[offset] = '/';
+	offset++;
+	for (int8_t i = 5; i >= 0; i--) {
+		hex2char(addr->a.val[i] >> 4, &path[offset]);
```
Our coding guidelines actually require that we check for error responses, so we probably need to do something like 
```suggestion
		err = hex2char(addr->a.val[i] >> 4, &path[offset]);
		if (err != 0) {
		    return -EINVAL;
		}
```

Interestingly, `bin2hex` does not apply this and would silently fail on invalid input

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +32,122 @@
+	 *  "bt/<key>"
+	 */
+	const size_t prefix_len = sizeof("bt") - 1;
+	const size_t key_len = key == NULL ? 0 : sizeof('/') + strlen(key);
```
```suggestion
	const size_t key_len = sizeof('/') + strlen(key);
```

`key` isn't optional in this function, right?

Thalley reviewed (changes requested):
Minor changes requested, otherwise LGTM. 

I'm not 100% on whether `str` or `mem` functions are ideal here, and I expect there's not a one true "best" solution

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +32,122 @@
+	const size_t subsys_len = sizeof('/') + strlen(subsys);
+	const size_t addr_str_len = sizeof('/') + sizeof("554433221100t") - 1;
+	const size_t key_len = key == NULL ? 0 : sizeof('/') + strlen(key);
+	const size_t total_len = prefix_len + subsys_len + addr_str_len + key_len + sizeof('\0');
```
Probably needs something like 

```suggestion
	const size_t prefix_len = sizeof("bt") - 1;
	const size_t subsys_len = strlen(subsys);
	const size_t addr_str_len = sizeof("554433221100t") - 1;
	const size_t key_len = key == NULL ? 0 : strlen(key);
	const size_t total_len = prefix_len + sizeof('/') + subsys_len + sizeof('/') + addr_str_len + key == NULL ? 0 : (sizeof('/') + key_len) + sizeof('\0');
```

To deal with the sonar issues. I guess we don't have any CI tests for `CONFIG_BT_SETTINGS_USE_PRINTK=n` to verify correctness of these functions?

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -33,75 +32,130 @@
+	 */
+	const size_t prefix_len = sizeof("bt") - 1;
+	const size_t subsys_len = strlen(subsys);
+	const size_t addr_str_len = sizeof("554433221100t") - 1;
```
Might be worth mentioning what `554433221100t` refers to (BD addr + type)

cvinayak committed:
tests: bsim: Bluetooth: Cover settings without snprintk use

Cover settings without snprintk use in bsim test.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/settings.c:
```suggestion
	const size_t prefix_len = sizeof("bt") - 1U;
	const size_t subsys_len = strlen(subsys);
	const size_t addr_str_len = sizeof("554433221100t") - 1U;
	const size_t key_len = (key == NULL) ? 0U : strlen(key);
	const size_t total_len = prefix_len +
				 sizeof('/') + subsys_len +
				 sizeof('/') + addr_str_len +
				 ((key == NULL) ? 0U : (sizeof('/') + key_len)) +
```

Minor (technically a coding guidelines violation)

cvinayak replied:
added `U`

Thalley commented on subsys/bluetooth/host/settings.c:
I recently learned that `sizeof('\0')` or `sizeof('/')` is actually the same as `sizeof(int)` :s 

So we can either do `sizeof(char)` or something like 

```c
const char null_term = '\0';
sizeof(null_term);
```

Weird C behavior IMO :( So sizeof constant literal chars cannot/should not be used

cvinayak replied:
explicit use of `/` or `\0`  is expressive as to what we are concatenating. keeping as-is.

Thalley replied:
I think we need to update it; the length check is otherwise invalid. 

`total_len` will be 12 larger than the actual length needed, which means that if `path_size is 50` and the _actual_ lenght is 40, then the `if (path_size < total_len) {` check will fail because the calculation for `total_len` will be 12 larger (52 instead of 40). So while `sizeof('/')` is nice for semantic reasons (make it clear that we are taking the size of this character), due to the way `sizeof` works, it will be wrong. 

I'm surprised this isn't caught by our CI, which also kind of exposes that our CI testing for these functions are missing some verification

Thalley replied:
One way that can work (just tested) is to make it `sizeof((char)'/')` which does return the expected size (1)

cvinayak replied:
Fixed, re-review.

cvinayak commented on subsys/bluetooth/host/settings.c:
cant figure where this line you refer to...

cvinayak reviewed (commented):
github changed,,,it is difficult to refer to older comments... please let me know if I missed some old ones

cvinayak commented:
> Sorry if this was already discussed and I missed it, but I'm worried that we'd now have two buildtime dependent branches of non-trivial code that will need to stay precisely in sync (generate & parse exactly the same strings). What would be the penalty of just forgetting the new Kconfig option and instead always use the snprintk-free variant of the implementation?

No worries. The code size are here: https://github.com/zephyrproject-rtos/zephyr/pull/100887#discussion_r2634389905

Use of `printk` in the said application increases the size by ~ 5KB, and using the non-snprintk variant increases it by a marginal size (this may vary more or less in the future, generally its more code to compile in tho).

This PR re-ensure `CONFIG_BT_SETTINGS_USE_PRINTK=n` truely does not use kernel's printk implementation, and hence reduce code size by ~ 2KB.

Thalley reviewed (approved):
No blockers :) LGTM
